public/calendar_reboot.js

```diff
@@ -865,6 +870,116 @@ Calendar.prototype.fillCalendar = function(monthObjectsArray) {
+        
+        //TODO extend this to support multiple callbacks
+        self.activityChangeFunctions.push(func);
+    };
+    
+        self.startWork = function() {
+        // Will increment the counter and possibly fire an event.
+        
+        self.activeCalls += 1;
+        console.log("starting work, number of active calls: " + self.activeCalls);
+        // Will dispatch the event backgroundActivityChange 
+        self.activityChangeFunctions.forEach(function(func) {
+            func(self.activeCalls);
+        });
+        
+    };
+    
+    self.endWork = function() {
+        // Will decrement the counter and maybe fire an event.
+        
+        
+        self.activeCalls -= 1;
+        console.log("ending work, number of active calls: " + self.activeCalls);
+        if (self.activeCalls < 0) {
+            console.error("No work has been started");
+        }
+        
+        // Dispatch the activityChanged listener
+        self.activityChangeFunctions.forEach(function(func) {
+            func(self.activeCalls);
+        });
+      
+    };
+    
+    self.getAllCalendarIds = function() {
+        // Returns a promise for the allCalendarIds object from storage
+        
+        var userId = self.user.currentUser.uid;
+        self.startWork();
+        return new Promise(function(resolve, reject) {
+            self.database.ref('/users/' + userId + '/checkit_allCalendarIds')
+            .once('value')  
+            .then(function(allCalendarIds) {
+                self.endWork();
+                if (allCalendarIds.val() === null) {
+                    reject("allCalendarIds not found");
+                }
+                else {
+                    return resolve(allCalendarIds.val());
+                }
+            })
+            .catch(function(err) {
+                self.endWork();
+                console.error(err);
+                return err;
+            })
+        })
+    };
+    
+    self.save = function(calendarObj) {
+        //save an App object (like a calendar object for example) in storage
+        
+        // Store the calendar state in the firebase database.
+        
+        
+        var userId = self.user.currentUser.uid;
+        var calUniqueId = calendarObj.state.uniqueId;
+        var calTitle = calendarObj.state.title;
+        
+        
+        // Store the user's allCalendarIds.
+        var updates = {};
+        updates['users/' + userId + '/checkit_allCalendarIds/' + calUniqueId] = calTitle;
+        
+        self.startWork();
+        return new Promise(function(resolve, reject) {
+        
+            self.getAllCalendarIds()
+                .then(function (allCalendarIds) {
+                    console.log("setting new value in database");
+                    firebase.database().ref().update(updates);
+                    console.log("DATA SET");
+                    self.endWork();
+                    resolve();  // At what point do I reject?
```

Comment:
> ` // At what point do I reject?`

You reject when there's an error. In this case, the only error that might happen will come from the firebase library. What happens when there's a firebase error? Which code gets executed?